### ISSUE TITLE
[STORM-132] Default Storm scheduler should evenly use resources when managing multiple topologies

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#3974]
<p><a href="https://github.com/nathanmarz/storm/issues/359" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/nathanmarz/storm/issues/359</a></p>

<p>Currently, a single topology is evenly spread across the cluster, but this is not the case for multiple topologies (it targets one node first, then the rest). The default scheduler should order the hosts for scheduling in terms of which has the least slots used.</p>

<p>----------<br/>
lyogavin: To confirm, we want to firstly look at the number of used slots on each host. Firstly try to balance the used slot number across all the hosts. For the assignment inside each host, we also try to evenly balance the usage of each slot. Am I understanding correctly?</p>

<p>When i coding, i realized it's actually a little complicated. Looks like there are many policies here we want to consider:<br/>
1. Evenness of resource usage. (Do we want to evaluate evenness according to number of slots used in each host or the number of executors? Maybe number of executors is better, but also make it very complicated)<br/>
2. Least rescheduling. We probabaly also want to make the assignment change as less as possible. Looks like this is why DefaultScheduler.bad-slots is coded the current way.<br/>
3. Number of workers.</p>

<p>Then the question is how do we prioritize those policies. Sometimes they conflict to each other. For example, sometimes the most even distribution may need the most reassignment.</p>

<p>Any thoughts?</p>

<p>----------<br/>
xumingming: @lyogavin I think you might have over-thought this, as @nathanmarz already confirmed, you just need to update EvenScheduler.sort-slots to make sure the slots in the least used node appear first in the available-slots list.</p>

<p>----------<br/>
lyogavin: Thanks James. I got what you mean. So i'll not worry about the usage of the executor usage. Only consider the balance of slots usage.</p>

<p>But I think just simply change the sort-slots to sort the slots based on slots usage may still not work too well. For example, let's say there are 2 hosts, with 10 slots on each. Say 1st one used 1 slots, 2nd one used 2 slots, and we want to assign another topology with 8 workers. If we simply sort the slots based on usage, we'll end up with the list <span class="error">&#91;0 0 0 0 0 0 0 0 1 1&#93;</span>, then the new assignment we'll get would be all from the 1st host. Not balanced.</p>

<p>So in the above pull request, I implemented a solution in the way similar to watershed algorithm. It would firstly pick the slots from least used host, until that host uses the same number of slots as the second least used slots. Then it evenly picks slots from the 2 least used hosts until reaches the 3rd one. Iterating this way, we can get the best balanced assignment.</p>

<p>What do you think?</p>

---
<details><summary><i>Originally reported by <a title="James Xu" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=xumingming">xumingming</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-132" target="_blank">Default Storm scheduler should evenly use resources when managing multiple topologies</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Kang Xiao" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=xiaokang">xiaokang</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2014-08-25T19:22:18+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user xiaokang opened a pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-132" title="Default Storm scheduler should evenly use resources when managing multiple topologies" class="issue-link" data-issue-key="STORM-132"><del>STORM-132</del></a> sort supervisor by free slot in desending order</p>

<p>    Sort supervisor by free slot in desending order in EvenScheduler to schedule more evenly between supervisor</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/xiaokang/incubator-storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/xiaokang/incubator-storm</a> <a href="https://issues.apache.org/jira/browse/STORM-132" title="Default Storm scheduler should evenly use resources when managing multiple topologies" class="issue-link" data-issue-key="STORM-132"><del>STORM-132</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #36</p>

<hr />
<p>commit 6b275d95fbdbc8374a215ecb2551f0fca3438d81<br/>
Author: Kang Xiao <kxiao.tiger@gmail.com><br/>
Date:   2014-02-18T15:23:50Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-132" title="Default Storm scheduler should evenly use resources when managing multiple topologies" class="issue-link" data-issue-key="STORM-132"><del>STORM-132</del></a> sort supervisor by free slot in desending order<br/>
    in EvenScheduler to schedule more evenly between supervisor</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-37130285" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-37130285</a></p>

<p>    If supervisor host1 has 2 free slots and host2 has 3 free slots, currently EvenScheduler will choose host1 for a topology requiring 1 worker. This PR is to make EvenScheduler choose host2 since it's more free than host1 to avoid load balance problem. Any suggestions on it?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r13728675" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r13728675</a></p>

<p>    &#8212; Diff: storm-core/src/clj/backtype/storm/scheduler/EvenScheduler.clj &#8212;<br/>
    @@ -22,7 +22,12 @@<br/>
 :implements <span class="error">&#91;backtype.storm.scheduler.IScheduler&#93;</span>))</p>

<p>     (defn sort-slots <span class="error">&#91;all-slots&#93;</span></p>
<ul class="alternate" type="square">
	<li>(let <span class="error">&#91;split-up (vals (group-by first all-slots))&#93;</span><br/>
    +  (let [split-up<br/>
    + (map second<br/>
    +   (reverse<br/>
    +     (sort<br/>
    +       (for [<span class="error">&#91;host ports&#93;</span> (group-by first all-slots)]<br/>
    + <span class="error">&#91;(count ports) ports&#93;</span>))))]
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Could we do something like this?<br/>
    ```<br/>
    (sort #(> (count %1) (count %2)) (vals (group-by first all-slots)))<br/>
    ```<br/>
    It would be nice to move this into its own function, so we can write a quick unit test for it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-48665643" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-48665643</a></p>

<p>    @xiaokang, we would like to get this in.  Could you take a look at the comment? Does that seem reasonable?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-49510624" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-49510624</a></p>

<p>    @d2r , thanks for your comment. It's reasonable and I'll update code.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-49514066" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-49514066</a></p>

<p>    @d2r , I'v updated the code using sort-by. </p>

<p>    I'd like to write a unit test. But I'm not sure which part of code should the unit test be against for, the whole sort-slots function or a separate function for (sort-by count (vals (group-by first all-slots)))?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-51512163" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-51512163</a></p>

<p>    @d2r , a unit test test-sort-slots is added for sort-slots function. By the way, (sort-by count  ...) in last commit get a reverse result so I changed it to the sort as you suggested.</p>

<p>    The new test-sort-slots runs ok but another case testing4j-test fails. I can not find any relationship between this test and the sort-slots code changes. I roll back the sort-slots and testing4j-test still fails, so it should be something wrong in the master branch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-51797563" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-51797563</a></p>

<p>    OK.  I liked your `sort-by` better than my suggestion.  Maybe `(sort-by count (complement compare) (vals (group-by first all-slots)))` would work better.</p>

<p>    I will take a look and see what might be the problem with the other tests.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r16066139" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r16066139</a></p>

<p>    &#8212; Diff: storm-core/test/clj/backtype/storm/scheduler_test.clj &#8212;<br/>
    @@ -259,3 +260,27 @@<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 3)))))<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 5)))))<br/>
 ))<br/>
    +<br/>
    +(deftest test-sort-slots<br/>
    +  ;; test supervisor2 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6702&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      ])))<br/>
    +  ;; test supervisor3 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6703&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      <span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor3" 6703&#93;</span><br/>
    +      ])))<br/>
    +  ;; test supervisor1 and supervisor2 has the same free slot<br/>
    +  (is (= '(<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>])))<br/>
    +    )<br/>
    &#8212; End diff &#8211;</p>

<p>    This test passes with or without your change.  Do we need this test?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-51810642" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-51810642</a></p>

<p>    > @d2r , a unit test test-sort-slots is added for sort-slots function. By the way, (sort-by count ...) in last commit get a reverse result so I changed it to the sort as you suggested.</p>

<p>    I think this will work (giving `>` function as the third argument).<br/>
    ```Clojure<br/>
    (sort-by count > (vals (group-by first all-slots)))<br/>
    ```</p>

<p>    > The new test-sort-slots runs ok but another case testing4j-test fails. I can not find any relationship between this test and the sort-slots code changes. I roll back the sort-slots and testing4j-test still fails, so it should be something wrong in the master branch.</p>

<p>    I tried running tests on master and on your branch, but I did not see any tests fail.</p>

<p>    I did notice that the final assertion in this PR passes even without any code change.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-52068360" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-52068360</a></p>

<p>    hi @d2r , I tired run tests on master and I also did not see any tests fail. However I find failure in the test reports file storm-core/target/test-reports/backtype.storm.scheduler-test.xml. Is there anything wrong with the test code?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-52073583" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-52073583</a></p>

<p>    "I did notice that the final assertion in this PR passes even without any code change."</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-52073945" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-52073945</a></p>

<p>    It's the reason that cause the test passed without any code change. I just change the assertion order of the test case and the case is failed without code change.</p>

<p>    The following code looks fine and it works. PR is updated.<br/>
      (sort-by count > (vals (group-by first all-slots)))</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-52083754" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-52083754</a></p>

<p>    > It's the reason that cause the test passed without any code change. I just change the assertion order of the test case and the case is failed without code change.</p>

<p>    OK, I think what is going on here is that the sorting is not stable, and so the test may pass or fail depending on things like which version of Java is in use (I know there was some change 7 <del>> 8 with un</del>/stable sorting.)</p>

<p>    I will make comments in-line.</p>



<p>    > I tired run tests on master and I also did not see any tests fail. However I find failure in the test reports file storm-core/target/test-reports/backtype.storm.scheduler-test.xml. Is there anything wrong with the test code?</p>


<p>    This could happen if the test report is left over from a previous run.  But I will double-check it.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r16191875" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r16191875</a></p>

<p>    &#8212; Diff: storm-core/test/clj/backtype/storm/scheduler_test.clj &#8212;<br/>
    @@ -259,3 +260,27 @@<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 3)))))<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 5)))))<br/>
 ))<br/>
    +<br/>
    +(deftest test-sort-slots<br/>
    +  ;; test supervisor1 and supervisor2 has the same free slot<br/>
    +  (is (= '(<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>])))<br/>
    &#8212; End diff &#8211;</p>

<p>    This assert can be removed, since unstable sorting could cause either of two valid orderings.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r16192380" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r16192380</a></p>

<p>    &#8212; Diff: storm-core/test/clj/backtype/storm/scheduler_test.clj &#8212;<br/>
    @@ -259,3 +260,27 @@<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 3)))))<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 5)))))<br/>
 ))<br/>
    +<br/>
    +(deftest test-sort-slots<br/>
    +  ;; test supervisor1 and supervisor2 has the same free slot<br/>
    +  (is (= '(<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>])))<br/>
    +  ;; test supervisor2 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6702&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      ])))<br/>
    +  ;; test supervisor3 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6703&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      <span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor3" 6703&#93;</span><br/>
    +      ])))<br/>
    +    )<br/>
    &#8212; End diff &#8211;</p>

<p>    For the above tests, are we sure the ports will always be returned in ascending order (e.g., 6700...6701...6702...)?</p>

<p>    If not, maybe we should just check the supervisor names and not compare the actual port numbers.</p>

<p>    If we are sure the port numbers will be in order, then these tests should be OK.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user d2r commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-52097241" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-52097241</a></p>

<p>    > I tired run tests on master and I also did not see any tests fail. However I find failure in the test reports file storm-core/target/test-reports/backtype.storm.scheduler-test.xml. Is there anything wrong with the test code?<br/>
    > <br/>
    > This could happen if the test report is left over from a previous run. But I will double-check it.</p>

<p>    I re-ran tests in a new working tree at commit 9968b8b03b9b8eeb9e2f9293d58460cb01647fb1 and I found no test failures under storm-core/target/test-reports.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r16298432" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r16298432</a></p>

<p>    &#8212; Diff: storm-core/test/clj/backtype/storm/scheduler_test.clj &#8212;<br/>
    @@ -259,3 +260,27 @@<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 3)))))<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 5)))))<br/>
 ))<br/>
    +<br/>
    +(deftest test-sort-slots<br/>
    +  ;; test supervisor1 and supervisor2 has the same free slot<br/>
    +  (is (= '(<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>])))<br/>
    &#8212; End diff &#8211;</p>

<p>    Yes, the order is not defined for supervisors with the same free slots.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user xiaokang commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#discussion_r16300197" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#discussion_r16300197</a></p>

<p>    &#8212; Diff: storm-core/test/clj/backtype/storm/scheduler_test.clj &#8212;<br/>
    @@ -259,3 +260,27 @@<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 3)))))<br/>
 (is (= false (.isSlotOccupied cluster (WorkerSlot. "supervisor1" (int 5)))))<br/>
 ))<br/>
    +<br/>
    +(deftest test-sort-slots<br/>
    +  ;; test supervisor1 and supervisor2 has the same free slot<br/>
    +  (is (= '(<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span>])))<br/>
    +  ;; test supervisor2 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor2" 6702&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      ])))<br/>
    +  ;; test supervisor3 has more free slots<br/>
    +  (is (= '(<span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6700&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +   <span class="error">&#91;"supervisor3" 6703&#93;</span>)<br/>
    + (sort-slots [<span class="error">&#91;"supervisor1" 6700&#93;</span> <span class="error">&#91;"supervisor1" 6701&#93;</span><br/>
    +      <span class="error">&#91;"supervisor2" 6700&#93;</span> <span class="error">&#91;"supervisor2" 6701&#93;</span> <span class="error">&#91;"supervisor2" 6702&#93;</span><br/>
    +      <span class="error">&#91;"supervisor3" 6700&#93;</span> <span class="error">&#91;"supervisor3" 6701&#93;</span> <span class="error">&#91;"supervisor3" 6702&#93;</span> <span class="error">&#91;"supervisor3" 6703&#93;</span><br/>
    +      ])))<br/>
    +    )<br/>
    &#8212; End diff &#8211;</p>

<p>    The port numbers will be in the same order as they are given in the arguments, since 'group-by' and 'vals' function both keep order of input collection or sequence. I also changed the port numbers order in the tests to make sure about this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user revans2 commented on the pull request:</p>

<p>    <a href="https://github.com/apache/incubator-storm/pull/36#issuecomment-53305759" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/incubator-storm/pull/36#issuecomment-53305759</a></p>

<p>    +1 looks good to me.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bbaugher">bbaugher</a>:</i>
<p>This is actually fixed in 0.9.3</p>
